README.md
# Flutter Stack Canvas

A simple canvas for placing widgets in a free style canvas-like manner by wrapping the widget in a Canvas Object and specifying its offset and size.

Furthermore, the canvas offers tranformation utilities to
- Scale the canvas: zoom in and out.
- Translate the canvas: move in the 4 directions.

All canvas tranformations can be animated.

See the example app for a comprehensive demo.

Inspired by [Rody Davis Jr](https://blog.codemagic.io/multi-touch-canvas-with-flutter/)

## Installation

Add `stack_canvas` as a dependency in your [pubspec.yaml](https://flutter.dev/docs/development/packages-and-plugins/using-packages) file.

Import Stack Canvas:
```dart
import 'package:stack_canvas:stack_canvas.dart;
```


## How to use

Simply use the `StackCanvas` widget to embed a new Canvas into your view. This is an empty canvas.
```dart
class MyView extends StatelessWidget {
    @override
    Widget build(BuildContext context) {
        return Container(
            StackCanvas(
                controller: StackCanvasController(),
            ),
        );
    }
}
```
| Empty Canvas |
| :----------: |
| ![Empty Canvas](https://raw.githubusercontent.com/Ali-Elganzory/stack_canvas/main/misc/empty.png) |



Should you need to add objects (obviously), use the `StackCanvasController`. Using this controller, adding, removing, and transforming the canvas becomes a simple task.
```dart
final StackCanvasController controller = StackCanvasController();
```
And give it to the `StackCanvas` you want to control.
```dart
StackCanvas(
    controller: controller,
),
```

Now, we can add any widgets we like by wrapping it inside a `CanvasObject`. `CanvasObject` is, in fact, a generic type and is defined as `CanvasObject<T>`; however, most of times, it will be `CanvasObject<Widget>`, for we're using flutter.
```dart
List<CanvasObject<Widget>> objects = [
    CanvasObject<Widget>(
        dx: 100,             // Offset in x-axis
        dy: 100,             // Offset in y-axis
        width: 80,
        height: 40,
        child: Container(    // The widget to be rendered
            color: Colors.red,
        )
    )
];
```
Ofcourse you can add multiple objects at the same time using that list.

Then add these widget objects to the canvas using the controller.
```dart
controller.addCanvasObjects(objects);
```

Voilà!
| A Widget on Canvas |
| :----------: |
| ![Single Widget](https://raw.githubusercontent.com/Ali-Elganzory/stack_canvas/main/misc/object.png) |


## Zoom & Move


### Zoom in & out

```dart
controller.zoomIn();
controller.zoomOut();
```

| Zoomed In | Zoomed Out |
| :-------: | :--------: |
| ![Zoomed In](https://raw.githubusercontent.com/Ali-Elganzory/stack_canvas/main/misc/zoomedin.png) | ![Zoomed Out](https://raw.githubusercontent.com/Ali-Elganzory/stack_canvas/main/misc/zoomedout.png) |


### Move in 4 directions
```dart
controller.moveUp();
controller.moveDown();
controller.moveLeft();
controller.moveRight();
```


## Customizing the Canvas

Here are all the customizable properties with their default values.
```dart
StackCanvas(
    width: double.maxFinite,      // Full width
    height: double.maxFinite,     // Full height
    backgroundColor: Colors.white,
    animationDuration: Duration(milliseconds: 400),
    controller: controller,
    disposeController: true,      // If set to false, you need to dispose the controller by yourself
)
```
```dart
StackCanvasController(
    zoomChangeUnit: 0.10,         // The speed of zooming  (scale factor)
    moveChangeUnit: 30.00,        // The speed of movement (translation value)
    offsetReference: Reference.TopLeft,
    zoomReference: Reference.TopLeft,
)
```

## Examples

### Example 1
| Canvas Transformations |
| :--------: |
| ![Transformation](https://raw.githubusercontent.com/Ali-Elganzory/stack_canvas/main/misc/transformation.gif) |

### Example 2
| Growing Triangles |
| :--------: |
| ![Growing Triangles](https://raw.githubusercontent.com/Ali-Elganzory/stack_canvas/main/misc/growing_triangles.gif) |



CHANGELOG.md
## [0.2.0] - 13 May 2021

### Added:
- Added Reference points for offsetting and scaling, separately.

### Removed:
- Residual code.


## [0.1.1] - 13 May 2021

### Added:
- Exposed scale and move speed parameters - customizable.


## [0.1.0] - 13 May 2021

### First Working Version
- Adding & removing widgets.
- Zooming in & out.
- Moving in the 4 directions.
- Some customizable canvas properties.



test/stack_canvas_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:stack_canvas/stack_canvas.dart';

void main() {
}


lib/View/stack_canvas.dart
import 'package:flutter/material.dart';

import 'package:stack_canvas/Controller/stack_canvas_controller.dart';
import 'package:stack_canvas/Models/canvas_object.dart';

class StackCanvas extends StatefulWidget {
  const StackCanvas({
    this.width = double.maxFinite,
    this.height = double.maxFinite,
    this.backgroundColor = Colors.white,
    this.animationDuration = const Duration(milliseconds: 400),
    required this.canvasController,
    this.disposeController = true,
  });

  final double width;
  final double height;
  final Color backgroundColor;
  final Duration animationDuration;
  final StackCanvasController canvasController;
  final bool disposeController;

  @override
  _StackCanvasState createState() => _StackCanvasState();
}

class _StackCanvasState extends State<StackCanvas> {
  final GlobalKey<State> containerKey = GlobalKey<State>();

  @override
  void dispose() {
    super.dispose();
    if (widget.disposeController) widget.canvasController.dispose();
  }

  List<CanvasObject<Widget>> canvasObjects = [];

  @override
  Widget build(BuildContext context) {
    return Container(
      key: containerKey,
      width: this.widget.width,
      height: this.widget.height,
      color: widget.backgroundColor,
      child: LayoutBuilder(
        builder: (_, constraints) {
          widget.canvasController.canvasSize = constraints.smallest;
          return StreamBuilder<List<CanvasObject<Widget>>>(
            stream: widget.canvasController.canvasObjectsStream,
            builder: (_, snapshot) {
              if (snapshot.hasData) canvasObjects = snapshot.data!;

              return Stack(
                children: [
                  for (final CanvasObject<Widget> obj in canvasObjects)
                    AnimatedPositioned.fromRect(
                      duration: widget.animationDuration,
                      rect: obj.rect,
                      child: obj.child,
                    ),
                ],
              );
            },
          );
        },
      ),
    );
  }
}


lib/Models/canvas_object.dart
import 'dart:ui';

import 'package:stack_canvas/Utils/rect_transform_ext.dart';

class CanvasObject<T> {
  const CanvasObject({
    required double dx,
    required double dy,
    required double width,
    required double height,
    required T child,
  })   : _dx = dx,
        _dy = dy,
        _width = width,
        _height = height,
        _child = child;

  final double _dx;
  final double _dy;
  final double _width;
  final double _height;
  final T _child;

  Offset get offset => Offset(_dx, _dy);
  Size get size => Size(_width, _height);
  Rect get rect => offset & size;
  T get child => _child;

  CanvasObject<T> copyWith({
    double? dx,
    double? dy,
    double? width,
    double? height,
    T? child,
  }) {
    return CanvasObject<T>(
      dx: dx ?? _dx,
      dy: dy ?? _dy,
      width: width ?? _width,
      height: height ?? _height,
      child: child ?? _child,
    );
  }

  CanvasObject<T> transform(Offset offset, double scale) {
    final Rect transformedRect = this.rect.transform(offset, scale);

    return this.copyWith(
      dx: transformedRect.left,
      dy: transformedRect.top,
      width: transformedRect.width,
      height: transformedRect.height,
    );
  }

  CanvasObject<T> transformFromOrigin(
      Offset offset, double scale, Offset origin) {
    final Rect transformedRect =
        this.rect.transformFromOrigin(offset, scale, origin);

    return this.copyWith(
      dx: transformedRect.left,
      dy: transformedRect.top,
      width: transformedRect.width,
      height: transformedRect.height,
    );
  }
}


lib/Utils/rect_transform_ext.dart
import 'dart:ui';

extension tranformations on Rect {
  Rect scaleInPlace(double scale) {
    final double width = this.width * scale;
    final double height = this.height * scale;
    return Rect.fromLTWH(this.left - (width - this.width) / 2,
        this.top - (height - this.height) / 2, width, height);
  }

  Rect scaleFromOrigin(double scale, Offset origin) {
    final double width = this.width * scale;
    final double height = this.height * scale;
    final Offset offset = origin - (origin - this.topLeft) * scale;
    return offset & Size(width, height);
  }

  Rect transform(Offset offset, double scale) {
    return this.shift(offset).scaleInPlace(scale);
  }

  Rect transformFromOrigin(Offset offset, double scale, Offset origin) {
    return this.shift(offset).scaleFromOrigin(scale, origin);
  }
}


lib/Controller/stack_canvas_controller.dart
import 'dart:async';
import 'dart:math';
import 'dart:ui';

import 'package:flutter/widgets.dart';
import 'package:stack_canvas/Models/canvas_object.dart';

enum Reference { TopLeft, Center }

class StackCanvasController {
  StackCanvasController({
    double zoomChangeUnit = 0.10,
    double moveChangeUnit = 30.00,
    Reference offsetReference = Reference.TopLeft,
    Reference zoomReference = Reference.TopLeft,
  })  : _scaleChangeUnit = zoomChangeUnit,
        _offsetChangeUnit = moveChangeUnit,
        offsetReference = offsetReference,
        zoomReference = zoomReference;

  // For updating the ui with widgets
  final StreamController<List<CanvasObject<Widget>>>
      _canvasObjectsStreamController =
      StreamController<List<CanvasObject<Widget>>>();
  Stream<List<CanvasObject<Widget>>> get canvasObjectsStream =>
      _canvasObjectsStreamController.stream;

  // Canvas size -- set by view
  Size _canvasSize = Size.zero;
  Size get canvasSize => _canvasSize;
  set canvasSize(Size size) {
    _canvasSize = size;
    _updateCanvas();
  }

  // Canvas transformation parameters
  Reference offsetReference;
  Reference zoomReference;
  static const Offset _defaultOffset = Offset.zero;
  static const double _defaultScale = 1;
  static const double _maxScaleMargin = 3;
  final double _scaleChangeUnit;
  final double _offsetChangeUnit;

  // Current canvas transformation values
  Offset _offset = _defaultOffset;
  Offset get offset => _offset;
  set offset(Offset offset) {
    _offset = offset;
    _updateCanvas();
  }

  double _scale = _defaultScale;
  double get scale => _scale;
  set scale(double scale) {
    _scale = scale;
    _updateCanvas();
  }

  // Transformation utilities
  void zoomIn() {
    scale = min(scale + _scaleChangeUnit, _maxScaleMargin);
  }

  void zoomOut() {
    scale = max(scale - _scaleChangeUnit, -_maxScaleMargin);
  }

  void moveRight() {
    offset = offset.translate(-_offsetChangeUnit, 0);
  }

  void moveLeft() {
    offset = offset.translate(_offsetChangeUnit, 0);
  }

  void moveUp() {
    offset = offset.translate(0, _offsetChangeUnit);
  }

  void moveDown() {
    offset = offset.translate(0, -_offsetChangeUnit);
  }

  Offset get originPoint => offsetReference == Reference.TopLeft
      ? Offset.zero
      : canvasSize.center(Offset.zero);

  Offset get zoomPoint => zoomReference == Reference.TopLeft
      ? Offset.zero
      : canvasSize.center(Offset.zero);

  // Canvas objects to be rendered
  final List<CanvasObject<Widget>> _canvasObjects = [];
  List<CanvasObject<Widget>> get canvasObjects {
    // Apply canvas transformations to the list
    return _canvasObjects
        .map((obj) =>
            obj.transformFromOrigin(originPoint + offset, scale, zoomPoint))
        .toList();
  }

  // Add canvas to be rendered
  void addCanvasObjects(List<CanvasObject<Widget>> objects) {
    _canvasObjects.addAll(objects);
    _updateCanvas();
  }

  // Remove canvas object
  void removeCanvasObject(CanvasObject<Widget> object) {
    _canvasObjects.remove(object);
    _updateCanvas();
  }

  // Publish the updated list of canvas objects
  void _updateCanvas() {
    _canvasObjectsStreamController.add(canvasObjects);
  }

  // Restore the canvas to its default transformation
  void resetCanvasTransformation() {
    _offset = _defaultOffset;
    _scale = _defaultScale;
    _updateCanvas();
  }

  void clearCanvas() {
    _canvasObjects.clear();
    _updateCanvas();
  }

  // Dispose any unmanaged resources
  void dispose() {
    _canvasObjectsStreamController.close();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:stack_canvas/stack_canvas.dart';

void main() {
  runApp(App());
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final StackCanvasController controller = StackCanvasController(
    offsetReference: Reference.Center,
    zoomReference: Reference.Center,
  );

  @override
  void initState() {
    super.initState();

    // Add dummy objects
    controller.addCanvasObjects([
      CanvasObject<Widget>(
        dx: -100,
        dy: -100,
        width: 40,
        height: 60,
        child: Container(
          color: Colors.red,
        ),
      ),
      CanvasObject<Widget>(
        dx: 30,
        dy: -120,
        width: 80,
        height: 80,
        child: Container(
          color: Colors.blue,
        ),
      ),
      CanvasObject<Widget>(
        dx: 10,
        dy: -20,
        width: 100,
        height: 60,
        child: Container(
          color: Colors.purple,
        ),
      ),
    ]);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Stack Canvas"),
        actions: [
          IconButton(
            icon: Icon(Icons.arrow_drop_up_rounded),
            tooltip: "Move upward",
            onPressed: () {
              controller.moveUp();
            },
          ),
          IconButton(
            icon: Icon(Icons.arrow_drop_down_rounded),
            tooltip: "Move downward",
            onPressed: () {
              controller.moveDown();
            },
          ),
          IconButton(
            icon: Icon(Icons.arrow_left_rounded),
            tooltip: "Move to the left",
            onPressed: () {
              controller.moveLeft();
            },
          ),
          IconButton(
            icon: Icon(Icons.arrow_right_rounded),
            tooltip: "Move to the right",
            onPressed: () {
              controller.moveRight();
            },
          ),
          IconButton(
            icon: Icon(Icons.zoom_in_rounded),
            tooltip: "Zoom in",
            onPressed: () {
              controller.zoomIn();
            },
          ),
          IconButton(
            icon: Icon(Icons.zoom_out_rounded),
            tooltip: "Zoom out",
            onPressed: () {
              controller.zoomOut();
            },
          ),
        ],
      ),
      body: Center(
        child: StackCanvas(
          width: double.maxFinite,
          height: double.maxFinite,
          canvasController: controller,
        ),
      ),
      floatingActionButton: FloatingActionButton(
        tooltip: 'Reset canvas transforamtion',
        child: Icon(Icons.refresh_rounded),
        onPressed: () {
          controller.resetCanvasTransformation();
        },
      ),
    );
  }
}
